nixos/modules/services/networking/spacecookie.nix

```diff
@@ -0,0 +1,90 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let cfg = config.services.spacecookie;
+    defaultUser = "spacecookie";
+    configFile = pkgs.writeText "spacecookie.json" (lib.generators.toJSON {} {
+        hostname = cfg.hostname;
+        port = cfg.port;
+        user = cfg.user;
+        root = cfg.root;
+      });
+in {
+
+  options = {
+
+    services.spacecookie = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Enable the Spacecookie Gopher Daemon";
+      };
+
+      hostname = mkOption {
+        type = types.str;
+        default = "localhost";
+        description = "The hostname the service is reachable via. Clients will use this hostname for further requests after loading the initial gopher menu.";
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 70;
+        description = "Port the gopher service should be exposed on.";
+      };
+
+      root = mkOption {
+        type = types.path;
+        default = "/srv/gopher";
+        description = "The root directory spacecookie serves via gopher.";
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "spacecookie";
+        description = "User the daemon should be run as.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    systemd.sockets.spacecookie = {
+      description = "Socket for the Spacecookie Gopher Server";
+      wantedBy = [ "sockets.target" ];
+      listenStreams = [ "[::]:${toString cfg.port}" ];
+      socketConfig.BindIPv6Only = "both";
+      socketConfig.SocketUser = cfg.user;
+      socketConfig.SocketGroup = cfg.user;
+    };
+
+    systemd.services.spacecookie = {
+      description = "Spacecookie Gopher Server";
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "spacecookie.socket" ];
+
+      preStart = ''
+        mkdir -p ${cfg.root}
+        chown ${cfg.user}:${cfg.user} ${cfg.root}
+      '';
+
+      serviceConfig = {
```

Comment:
Why does this need to run as root? Given you are using a socket I would imagine the application should probably run as an unprivileged user. Just a guess.

Reply:
Probably also worth adding some `systemd` magic to this to lock it down a bit further.

Reply:
Yeah, it should! Missed that, although the service would switch to the configured user by itself. What do you have in mind except for setting `User=`?

Reply:
Please also set `Group`.

As far as `systemd` magic... I bet @dasJ or @flokli have some great ideas.

Reply:
Added `Group` as well.

Reply:
Okay sooo. Instead of using a static user, you can also use a [DynamicUser](http://0pointer.net/blog/dynamic-users-with-systemd.html) which would allocate a user if the unit is started.
This also means a lot of other sandboxing options are enabled (like `ProtectSystem` which mounts most of your system read-only) (see also [here](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#DynamicUser=)). Using a dynamic user would mean you don't have to define a user below.

To keep it short, `DynamicUser = true;` implies:
```
ProtectSystem = "strict";
ProtectHome = "read-only";
PrivateTmp = true;
```

In order to get a lot more sandboxing, I also recommend (although I didn't test it for spacecookie):
```
ProtectHome = true; # Stricter than "read-only" which is implied by `DynamicUser`
PrivateDevices = true;
ProtectKernelTunables = true;
ProtectKernelModules = true;
ProtectControlGroups = true;

DynamicUser = true;

CapabilityBoundingSet = "";
NoNewPrivileges = true;

LockPersonality = true;
RestrictRealtime = true;
PrivateMounts = true;
PrivateUsers = true;
# If you are brave: SystemCallFilter = "";
SystemCallArchitectures = "native";
RestrictAddressFamilies = "AF_INET AF_INET6";
```

But I totally understand if that is too much, using DynamicUser or at least ProtectSystem=strict is probably enough.

Reply:
I'd love to use `DynamicUser`, but how can I setup a socket for that user beforehand using systemd? Systemd sockets don't seem to have a comparable option...

Reply:
You can set a `User` and `Group` value with `DynamicUser`. I wonder if that would work or not... it does with `mysql` iirc.

Reply:
> I'd love to use DynamicUser, but how can I setup a socket for that user beforehand using systemd? Systemd sockets don't seem to have a comparable option...

I'm not sure I get what you're missing from systemd sockets, and how spacecookie itself handles connections. Can you elaborate?

Reply:
> You can set a User and Group value with DynamicUser. I wonder if that would work or not... it does with mysql iirc.

That does not work, as `spacecookie.socket` is set up before the user and group which is only created, when the service is actually started.

This results in the following error for `spacecookie.socket`:

```
Dec 10 14:04:50 wolfgang systemd[1]: Starting Socket for the Spacecookie Gopher Server.
Dec 10 14:04:50 wolfgang systemd[2372]: spacecookie.socket: Failed to resolve user spacecookie: No such process
Dec 10 14:04:50 wolfgang systemd[1]: spacecookie.socket: Control process exited, code=exited, status=217/USER
Dec 10 14:04:50 wolfgang systemd[1]: spacecookie.socket: Failed with result 'exit-code'.
Dec 10 14:04:50 wolfgang systemd[1]: Failed to listen on Socket for the Spacecookie Gopher Server.
```

Reply:
> I'm not sure I get what you're missing from systemd sockets, and how spacecookie itself handles connections. Can you elaborate?

Well first of all, if systemd sets up the socket, the service doesn't have to be started as `root` to bind to its well-known port, which is, why it'd be best to use the systemd socket approach. Also systemd can start the service on-demand (socket activation), if we use this method.

Reply:
@aanderse Did you mean `services.mysql`? I had a look at that one and it does not use systemd sockets.

Reply:
No, not `systemd` sockets but `mysql` provisions the socket. Makes sense that it doesn't work, now that I think about it more. Thanks for checking, at least!